src/middleware/auto-webp.ts

```diff
@@ -0,0 +1,150 @@
+
+export interface ImageServiceOptions {
+  rootPath?: string,
+  webp?: boolean
+}
+
+export type FetchFn = (req: RequestInfo, init?: RequestInit) => Promise<Response>
+
+export function imageService(origin: FetchFn | string): FetchFn {
+  return async function imageServiceFetch(req: RequestInfo, init?: RequestInit) {
+    const parser = defaultParser
+    if (typeof origin === "string") {
+      origin = proxy(origin)
+    }
+    if (typeof req === "string") req = new Request(req, init)
+
+    const op = parser(new URL(req.url))
+    // check if client supports webp
+    const webp = webpAllowed(req)
+
+    // generate a cache key
+    const key = cacheKey(op, webp)
+    let resp: Response = await get(key)
+    if(resp){
+      resp.headers.set("Fly-Cache", "HIT")
+      return resp
+    }
+
+    const breq = new Request(op.url.toString(), req)
+    resp = await fetchFromCache(breq, origin)
+
+    let start = Date.now()
+
+    // check if the req is a png or jpeg image
+    if(!isImage(resp)) {
+      return resp
+    }
+
+    if (req.method === "GET"){
+      let img = await loadImage(resp)
+
+      if(webp){
+        img = img.webp({ force: true })
+        resp.headers.set("content-type", "image/webp")
+      }
+
+      const body = await img.toBuffer()
+      resp = new Response(body.data, resp)
+      resp.headers.set("content-length", body.data.byteLength.toString())
+
+      // set image in responseCache
+      await set(key, resp, { tags: [op.url.toString()], ttl: 3600 })
+      resp.headers.set("Fly-Cache", "MISS")
+    }
+    return resp
+  }
+}
+
+async function fetchFromCache(req: Request, origin: FetchFn) {
+  let start = Date.now()
+  let resp: Response = await get(req.url)
+  if (resp) {
+    resp.headers.set("Fly-Cache", "hit")
+    console.log(`Image fetch from cache (${Date.now() - start}):`, req.url)
+    return resp
+  }
+
+  resp = await origin(req)
+
+  if (resp.status === 200 && req.method === "GET") {
+    console.log(`Image fetch from URL (${Date.now() - start}):`, req.url)
+    start = Date.now()
+    await set(req.url, resp, { tags: [req.url], ttl: 3600 * 24 * 30 * 6 })
+    console.log(`Image write to cache (${Date.now() - start}):`, req.url, resp.headers.get("content-length"))
+    resp.headers.set("Fly-Cache", "miss")
+    return resp
+  }
+  return resp
+}
+
+async function loadImage(resp: Response): Promise<Image> {
+  if (!isImage(resp)) {
+    throw new Error("Response wasn't an image")
+  }
+  const raw = await resp.arrayBuffer()
+  const img = new Image(raw)
+  
+  const meta = img.metadata()
+  console.log("Image:", meta)
+  
+  return img
+}
+
+function isImage(resp: Response): boolean{
+  const contentType = resp.headers.get("Content-Type") || ""
+  if (contentType.includes("image/png") || contentType.includes("image/jpeg")) {
+    return true
+  }
+  return false
+}
+
+export interface TransformURL {
+  url: URL
+}
+
+export type TransformURLParser = (url: URL, opts?: ImageServiceOptions) => TransformURL
```

Comment:
I don't think the various URL transforms are necessary here, since all it's doing is webp.